src/app/reporting/Engine.cpp

```diff
@@ -336,6 +377,78 @@ void Engine::Run()
         mCurReadHandlerIdx = (mCurReadHandlerIdx + 1) % CHIP_IM_MAX_NUM_READ_HANDLER;
         readHandler        = imEngine->mReadHandlers + mCurReadHandlerIdx;
     }
+    for (auto & path : mDirtyPathList)
+    {
+        path.ClearDirty();
+    }
+    return;
+}
+
+void Engine::SetDirty(ClusterInfo & aClusterInfo)
+{
+    InteractionModelEngine * imEngine = InteractionModelEngine::GetInstance();
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    ScopedLock lock(*this);
+#endif // !CHIP_SYSTEM_CONFIG_NO_LOCKING
+
+    for (auto & path : mDirtyPathList)
+    {
+        if (path.IsDirty())
+        {
+            //If aClusterInfo is part of existing dirty path, skip it
+            //If aClusterInfo is super set of one existing path, update that path with new path.
+            if (path.IsAttributePathSupersetOf(aClusterInfo)) {
+                return;
+            }
+            if (aClusterInfo.IsAttributePathSupersetOf(path)) {
+                path.mListIndex = aClusterInfo.mListIndex;
+                path.mFieldId = aClusterInfo.mFieldId;
+                path.mFlags = aClusterInfo.mFlags;
+                return;
+            }
+        }
+    }
+
+    for (auto & readHandler : imEngine->mReadHandlers)
+    {
+        if (readHandler.IsSubscription() && readHandler.IsReportable())
+        {
+            ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
+
+            while (clusterInstance != nullptr)
```

Comment:
Why not for loop here?

Reply:
For loop is safer, what if some one added a "continue" in the loop in the future?